internal/restic/snapshot_policy.go

```diff
@@ -180,6 +217,20 @@ func ApplyPolicy(list Snapshots, p ExpirePolicy) (keep, remove Snapshots, reason
 		{p.Yearly, y, -1, "yearly snapshot"},
 	}
 
+	// These buckets are for keeping snapshots of given type within duration
+	var bucketsWithin = [5]struct {
+		Within Duration
+		bucker func(d time.Time, nr int) int
+		Last   int
+		reason string
+	}{
+		{p.HourlyWithin, ymdh, -1, "hourly within"},
+		{p.DailyWithin, ymd, -1, "daily within"},
+		{p.WeeklyWithin, yw, -1, "weekly within"},
+		{p.MonthlyWithin, ym, -1, "monthly within"},
+		{p.YearlyWithin, y, -1, "yearly within"},
+	}
+
 	latest := findLatestTimestamp(list)
```

Comment:
I wonder whether we should bound that timestamp to the current time. Or probably even better ignore all snapshots from the future for this calculation (unless there's only a snapshot). Otherwise you could wipe most snapshots by just creating a snapshot with a timestamp far into the future.

Reply:
Should perhaps timestamps in the future be an error for all policy-based forgets (both these new switches and the pre-existing one)? Seems like that indicates a repo in a pretty bent shape.

Reply:
I like the idea in general, but I'm not sure which corner cases could appear. Timestamps that are only slightly in the future (e.g. just a few minutes) for example are probably ok and quite likely caused by a not properly synchronized clock.

Reply:
Yes, an error might be a little strict. So how about issuing a warning (wouldn't `Warnf()` be the right way to go?) if any future-dated snapshots are found, before removing them?

My sense is that the best way to do this check would  at the start of ApplyPolicy, before any other action. At least before checking if `len(list)==0` and returning. This would mean that it would not only apply to the new parameters, but to all policy-based forgets. But I think that would nevertheless be much cleaner, and furthermore, incorrectly dated snapshots could easily mess with all the other time-based policies as well. Obviously they affect `--keep-within`, but `--keep-latest` would also be heavily affected, and having several incorrectly dated snapshots would also mean that something like `--keep-daily 3` would end up forgetting all new (correctly dated) backups.

The place I'm suggesting implementing this is here:

```go
func ApplyPolicy(list Snapshots, p ExpirePolicy) (keep, remove Snapshots, reasons []KeepReason) {
	sort.Sort(list)

// CHECK FOR ANY FUTURE SNAPSHOTS HERE, REMOVE THEM FROM list IF FOUND,
// AND ISSUE A WARNING.

	if p.Empty() {
		for _, sn := range list {
			reasons = append(reasons, KeepReason{
				Snapshot: sn,
				Matches:  []string{"policy is empty"},
			})
		}
		return list, remove, reasons
	}

	if len(list) == 0 {
		return list, nil, nil
	}
```

Reply:
As the policies are applied per host, I'd expect all snapshots of that host to be affected. And thus a wrong date would cancel itself out. That should prevent a skewed clock from breaking the forget policies. If someone accidentally creates a snapshot with a wrong date, it's effect would be limited to just one of the snapshots kept by `--keep-latest` and `--keep-daily`. 

Whereas for `--keep-within` it could cause removing all snapshots except the one with the wrong date. Therefore, I'd prefer only to slightly change the behavior of `--keep-within*` but not for the other options. That would basically just mean to slightly tweak `findLatestTimestamp`.

Regarding a warning I don't have a feeling for how much noise that could generate. I think we should start without a warning for now and maybe add it if this leads to too many complaints.

Reply:
OK, I've done the following:

  * Squashed the commits implementing the new switches.
  * Changed `findLatestTimestamp` so that it returns the latest timestamp that is in the past, ignoring any future timestamps.
  * I've made sure that the `snapshot_policy_test.go` tests run with this latest change (but one should note that the test repo does not include any future snapshots).

What I did NOT do:

  * I did not add any error or warning on forgetting from a repo which has future snapshots (regardless).
  * I did not add any future snapshots to the unit test repo. I thought that was too much of a change for what is hopefully an absolute edge case.
  * I did not squash the `findLatestTimestamp` changes into the other changes, since it is logically separate from implementing the new switches, and I think it is better to keep it visible.

Thoughts on this warning thing:

  * After thinking about it more, I now feel that the best place to put in a test+warning for future snapshots would be in `restic check`. A warning could have negative implications for automatic runs of commands such as `forget`, but when running `check`, one is presumably doing it in order to examine the output. If we find that repos with future snapshots are common in the wild, we will start hearing about it in the forum and that would justify a deeper look into whether it is likely that future snapshots could cause bad things to happen.
  * Should I file an issue to do this?

Reply:
FWIW, I think that reasoning regarding the warning for future snapshots makes a lot of sense, `check` is a much better place indeed. Good thinking!